Actions/Github-Helper.psm1

```diff
@@ -720,38 +720,74 @@ function GetArtifacts {
     )
 
     $headers = GetHeader -token $token
-    $allArtifacts = @()
-    $per_page = 100
-    $page = 1
+    $total_count = 0
     if ($version -eq 'latest') { $version = '*' }
-    Write-Host "Analyzing artifacts"
-    do {
-        $uri = "$api_url/repos/$repository/actions/artifacts?per_page=$($per_page)&page=$($page)"
+    # Download all artifacts matching branch and version
+    # We might have results from multiple workflow runs, but we will have all artifacts from the workflow run that created the first matching artifact
+    # Use the buildOutput artifact to determine the workflow run id (as that will always be there)
+    $artifactPattern = "*-$branch-*-$version"
+    # Use buildOutput artifact to determine the workflow run id to avoid excessive API calls
+    # Reason: A project called xx-main will match the artifact pattern *-main-*-version, and there might not be any artifacts matching the mask
+    $buildOutputPattern = "*-$branch-BuildOutput-$version"
+    Write-Host "Analyzing artifacts matching $artifactPattern"
+    while ($true) {
+        if ($total_count -eq 0) {
+            # First iteration - initialize variables
+            $matchingArtifacts = @()
+            $buildOutputArtifacts = @()
+            $per_page = 100
+            $page_no = 1
+        }
+        $uri = "$api_url/repos/$repository/actions/artifacts?per_page=$($per_page)&page=$($page_no)"
         Write-Host $uri
-        $artifactsJson = InvokeWebRequest -Headers $headers -Uri $uri
-        $artifacts = $artifactsJson | ConvertFrom-Json
-        $page++
-        $artifactPattern = "*-$branch-$mask-$version"
-        Write-Host "ArtifactPattern: $artifactPattern"
-        $allArtifacts += @($artifacts.artifacts | Where-Object { !$_.expired -and $_.name -like $artifactPattern })
-        $result = @()
-        $allArtifactsFound = $true
+        $artifacts = InvokeWebRequest -Headers $headers -Uri $uri | ConvertFrom-Json
+        # If no artifacts are read, we are done
+        if ($artifacts.artifacts.Count -eq 0) {
+            break
+        }
+        if ($total_count -eq 0) {
+            $total_count = $artifacts.total_count
+        }
+        elseif ($total_count -ne $artifacts.total_count) {
+            # The total count changed, restart the loop
+            $total_count = 0
+            continue
```

Comment:
Does this mean more artifacts were added since the previous read?

This would probably be fine for a single build job, but isn't it problematic for multiple build job as some come be 15+ minutes apart?
That would mean some build job use different "baseline" that other from the same workflow.

Reply:
No, this means that the count changed in the middle of this loop - another workflow might be pushing artifacts while we are scanning.